working/macros/feature-specification.md

```diff
@@ -352,16 +357,20 @@ separating declarations.
 In the future, we may decide to run `dart format` or some other lighter weight
 formatter on augmentations which would also enforce consistent whitespace.
 
+#### Rule #4: New types are declared separately from their augmentations
+
+If a macro declares a new type and then later augments it, this will result in
+separate type declarations. One normal one followed by an augmentation of that
```

Comment:
Is this to support augmenting definitions?
```dart
class A {
  void foo() {
    print('foo');
  }
}

augment class A {
  augment void foo() {
    print('before1');
    super augment foo();
    print('after1');
  }
}

augment class A {
  augment void foo() {
    print('before2');
    super augment foo();
    print('after2');
  }
}
```

Or the "last statement" as I referred to it before, is about this?
But we can put all augmentations sequentially like this.
```dart
class A {
  void foo() {
    print('foo');
  }

  augment void foo() {
    print('before1');
    super augment foo();
    print('after1');
  }

  augment void foo() {
    print('before2');
    super augment foo();
    print('after2');
  }
}
```
Ah, I see the issue - we cannot have non-augmentation `class A` and augmentation methods in it.
So, we do them separately.
```dart
class A {
  void foo() {
    print('foo');
  }
}

augment class A {
  augment void foo() {
    print('before1');
    super augment foo();
    print('after1');
  }

  augment void foo() {
    print('before2');
    super augment foo();
    print('after2');
  }
}
```

This seems fine to me - most macros probably will generate new methods, not augment definitions.

But if we generate a new class, and just add new methods to it - currently we also call it "augmentation", right? If we have
```dart
@MakeClassB
class A {}
```
then
```dart
library augment 'a.macro.dart';

@AddMethodFoo
class B {}
```
Do we get this?
```dart
library augment 'a.macro.dart';

@AddMethodFoo
class B {
  void foo() {}
}
```
or this?
```dart
library augment 'a.macro.dart';

@AddMethodFoo
class B {
  void foo() {}
}

augment class B {
  void foo() {}
}
```

Reply:
You get the latter, which is what this paragraph is attempting to describe :).

```dart
library augment 'a.macro.dart';

@AddMethodFoo
class B {
  void foo() {}
}

augment class B {
  void foo() {}
}
```

Reply:
Oh, sorry, my last example was not correct.
I did not mean to have two `foo()` declarations.
So, it should be instead
```dart
library augment 'a.macro.dart';

@AddMethodFoo
class B {}

augment class B {
  void foo() {}
}
```

Which seems unfortunate. I can imagine a macro that is applied to a function for example, to create a class that wraps all parameters, and the created class itself is annotated with `@generateEquals`. It would seem natural to put the generated `operator ==()` into the same class.

Reply:
> Which seems unfortunate. I can imagine a macro that is applied to a function for example, to create a class that wraps all parameters, and the created class itself is annotated with `@generateEquals`. It would seem natural to put the generated `operator ==()` into the same class.

I agree it is less than ideal from a user perspective. I have it this way because it is the easiest/safest way to generate the library itself without actually parsing the code. I could explicitly call that out though?

Reply:
Sorry, I don't understand. What parsing?

Reply:
> Sorry, I don't understand. What parsing?

When you declare types in phase one, what we get is mostly just an opaque String with some identifiers in it. If we wanted to merge _new_ declarations from phase 2 into that declaration, we would have to parse it to find out where in the string we could inject new declarations. With the current model we don't have to do that, because we get the individual augmentations for a given type as standalone declarations and synthesize the surrounding `augment <type> {...}` declaration. So we can easily group them together without parsing.